File: apps/api/app/schemas/custom_agents.py
"""Schemas for the custom agent builder system."""

from datetime import datetime
from enum import StrEnum
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field


# ── Enums ────────────────────────────────────────────────────────────────────

class AgentStatus(StrEnum):
    DRAFT = "draft"
    ACTIVE = "active"
    PAUSED = "paused"
    ARCHIVED = "archived"


class ToolPermissionLevel(StrEnum):
    READ = "read"
    WRITE = "write"
    EXECUTE = "execute"
    ADMIN = "admin"


class AgentModelProvider(StrEnum):
    OPENAI = "openai"
    ANTHROPIC = "anthropic"
    DEEPSEEK = "deepseek"
    GOOGLE = "google"
    LOCAL = "local"


class OutputFormat(StrEnum):
    TEXT = "text"
    JSON = "json"
    MARKDOWN = "markdown"
    TABLE = "table"
    CODE = "code"
    EMAIL = "email"


class WorkflowType(StrEnum):
    SEQUENTIAL = "sequential"
    PARALLEL = "parallel"
    SUPERVISOR = "supervisor"
    CUSTOM = "custom"


class MemoryType(StrEnum):
    PREFERENCE = "preference"
    FACT = "fact"
    DECISION = "decision"
    CONTEXT = "context"
    USER_INFO = "user_info"
    PROJECT = "project"


class RunStatus(StrEnum):
    QUEUED = "queued"
    RUNNING = "running"
    PAUSED = "paused"
    COMPLETED = "completed"
    FAILED = "failed"
    CANCELLED = "cancelled"
    WAITING_APPROVAL = "waiting_approval"


class StepType(StrEnum):
    PLANNER = "planner"
    RESEARCH = "research"
    TOOL_CALL = "tool_call"
    KNOWLEDGE_RETRIEVAL = "knowledge_retrieval"
    LLM_CALL = "llm_call"
    SUB_AGENT = "sub_agent"
    REVIEWER = "reviewer"
    OUTPUT_FORMAT = "output_format"
    APPROVAL_CHECK = "approval_check"


class KnowledgeBaseStatus(StrEnum):
    PENDING = "pending"
    PROCESSING = "processing"
    READY = "ready"
    ERROR = "error"


# ── Custom Agent ─────────────────────────────────────────────────────────────

class CustomAgentBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=200)
    description: str | None = Field(default=None, max_length=2000)
    role_instructions: str | None = Field(default=None, max_length=8000)
    model_provider: AgentModelProvider = AgentModelProvider.DEEPSEEK
    model_name: str = "deepseek-chat"
    temperature: float = Field(default=0.7, ge=0, le=2)
    max_tokens: int = Field(default=4096, ge=1, le=128000)
    memory_enabled: bool = False
    memory_ttl_days: int = Field(default=30, ge=1, le=365)
    output_format: OutputFormat = OutputFormat.TEXT
    output_schema: dict | None = None
    workflow_type: WorkflowType = WorkflowType.SEQUENTIAL
    workflow_graph: dict | None = None
    status: AgentStatus = AgentStatus.DRAFT
    metadata_json: dict = Field(default_factory=dict)


class CustomAgentCreate(CustomAgentBase):
    workspace_id: UUID
    created_by: UUID | None = None


class CustomAgentUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=200)
    description: str | None = Field(default=None, max_length=2000)
    role_instructions: str | None = Field(default=None, max_length=8000)
    model_provider: AgentModelProvider | None = None
    model_name: str | None = None
    temperature: float | None = Field(default=None, ge=0, le=2)
    max_tokens: int | None = Field(default=None, ge=1, le=128000)
    memory_enabled: bool | None = None
    memory_ttl_days: int | None = Field(default=None, ge=1, le=365)
    output_format: OutputFormat | None = None
    output_schema: dict | None = None
    workflow_type: WorkflowType | None = None
    workflow_graph: dict | None = None
    status: AgentStatus | None = None
    metadata_json: dict | None = None
    updated_by: UUID | None = None


class CustomAgentResponse(CustomAgentBase):
    id: UUID
    workspace_id: UUID
    created_by: UUID | None = None
    updated_by: UUID | None = None
    created_at: datetime
    updated_at: datetime
    tools: list["AgentToolResponse"] = Field(default_factory=list)
    sub_agents: list["SubAgentResponse"] = Field(default_factory=list)
    knowledge_base_ids: list[UUID] = Field(default_factory=list)

    model_config = ConfigDict(from_attributes=True)


class CustomAgentSummary(BaseModel):
    id: UUID
    workspace_id: UUID
    name: str
    description: str | None
    model_provider: AgentModelProvider
    model_name: str
    memory_enabled: bool
    status: AgentStatus
    created_at: datetime
    updated_at: datetime
    tool_count: int = 0
    sub_agent_count: int = 0
    thread_count: int = 0

    model_config = ConfigDict(from_attributes=True)


# ── Agent Tools ──────────────────────────────────────────────────────────────

class AgentToolBase(BaseModel):
    tool_name: str = Field(..., min_length=1, max_length=100)
    tool_config: dict = Field(default_factory=dict)
    enabled: bool = True
    permission_level: ToolPermissionLevel = ToolPermissionLevel.READ
    requires_approval: bool = False
    rate_limit_per_day: int | None = None
    allowed_domains: list[str] | None = None
    allowed_actions: list[str] | None = None


class AgentToolCreate(AgentToolBase):
    agent_id: UUID


class AgentToolUpdate(BaseModel):
    tool_config: dict | None = None
    enabled: bool | None = None
    permission_level: ToolPermissionLevel | None = None
    requires_approval: bool | None = None
    rate_limit_per_day: int | None = None
    allowed_domains: list[str] | None = None
    allowed_actions: list[str] | None = None


class AgentToolResponse(AgentToolBase):
    id: UUID
    workspace_id: UUID
    agent_id: UUID
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


# ── Knowledge Bases ──────────────────────────────────────────────────────────

class KnowledgeBaseBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=200)
    description: str | None = Field(default=None, max_length=2000)
    source_type: str = "upload"
    embedding_model: str = "text-embedding-3-small"
    embedding_provider: str = "openai"


class KnowledgeBaseCreate(KnowledgeBaseBase):
    workspace_id: UUID
    created_by: UUID | None = None


class KnowledgeBaseUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=200)
    description: str | None = Field(default=None, max_length=2000)


class KnowledgeBaseResponse(KnowledgeBaseBase):
    id: UUID
    workspace_id: UUID
    file_count: int = 0
    chunk_count: int = 0
    status: KnowledgeBaseStatus
    created_by: UUID | None = None
    created_at: datetime
    updated_at: datetime
    files: list["KnowledgeBaseFileResponse"] = Field(default_factory=list)

    model_config = ConfigDict(from_attributes=True)


class KnowledgeBaseFileResponse(BaseModel):
    id: UUID
    knowledge_base_id: UUID
    file_name: str
    file_path: str
    file_type: str
    file_size_bytes: int | None = None
    chunk_count: int = 0
    status: KnowledgeBaseStatus
    error_message: str | None = None
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


class AgentKnowledgeBaseLink(BaseModel):
    agent_id: UUID
    knowledge_base_id: UUID
    retrieval_priority: int = Field(default=1, ge=1, le=10)
    max_chunks_per_query: int = Field(default=5, ge=1, le=50)
    similarity_threshold: float = Field(default=0.75, ge=0, le=1)


# ── Sub-Agents ───────────────────────────────────────────────────────────────

class SubAgentBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=200)
    role: str = Field(..., min_length=1, max_length=100)
    instructions: str = Field(..., min_length=1, max_length=8000)
    model_provider: AgentModelProvider | None = None
    model_name: str | None = None
    tools_json: list[str] = Field(default_factory=list)
    execution_order: int = Field(default=1, ge=1)
    enabled: bool = True
    requires_approval: bool = False


class SubAgentCreate(SubAgentBase):
    parent_agent_id: UUID


class SubAgentUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=200)
    role: str | None = Field(default=None, min_length=1, max_length=100)
    instructions: str | None = Field(default=None, max_length=8000)
    model_provider: AgentModelProvider | None = None
    model_name: str | None = None
    tools_json: list[str] | None = None
    execution_order: int | None = Field(default=None, ge=1)
    enabled: bool | None = None
    requires_approval: bool | None = None


class SubAgentResponse(SubAgentBase):
    id: UUID
    workspace_id: UUID
    parent_agent_id: UUID
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


# ── Agent Threads ────────────────────────────────────────────────────────────

class AgentThreadCreate(BaseModel):
    agent_id: UUID
    title: str | None = Field(default=None, max_length=500)
    created_by: UUID | None = None
    metadata_json: dict = Field(default_factory=dict)


class AgentThreadUpdate(BaseModel):
    title: str | None = Field(default=None, max_length=500)
    status: str | None = None


class AgentThreadResponse(BaseModel):
    id: UUID
    workspace_id: UUID
    agent_id: UUID
    title: str | None
    status: str
    metadata_json: dict
    created_by: UUID | None
    created_at: datetime
    updated_at: datetime
    message_count: int = 0
    last_message_at: datetime | None = None

    model_config = ConfigDict(from_attributes=True)


# ── Messages ─────────────────────────────────────────────────────────────────

class AgentMessageCreate(BaseModel):
    thread_id: UUID
    agent_id: UUID | None = None
    role: str = Field(..., pattern=r"^(user|assistant|system|tool|sub_agent)$")
    content: str | None = None
    tool_calls_json: list[dict] | None = None
    tool_call_id: str | None = None
    sub_agent_name: str | None = None
    metadata_json: dict = Field(default_factory=dict)


class AgentMessageResponse(BaseModel):
    id: UUID
    workspace_id: UUID
    thread_id: UUID
    agent_id: UUID | None
    role: str
    content: str | None
    tool_calls_json: list[dict] | None = None
    tool_call_id: str | None = None
    sub_agent_name: str | None = None
    token_count: int | None = None
    metadata_json: dict
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


# ── Agent Memories ───────────────────────────────────────────────────────────

class AgentMemoryCreate(BaseModel):
    agent_id: UUID
    thread_id: UUID | None = None
    memory_type: MemoryType = MemoryType.PREFERENCE
    content: str = Field(..., min_length=1, max_length=5000)
    importance: float = Field(default=0.5, ge=0, le=1)
    expires_at: datetime | None = None
    metadata_json: dict = Field(default_factory=dict)


class AgentMemoryUpdate(BaseModel):
    content: str | None = Field(default=None, min_length=1, max_length=5000)
    importance: float | None = Field(default=None, ge=0, le=1)
    metadata_json: dict | None = None


class AgentMemoryResponse(BaseModel):
    id: UUID
    workspace_id: UUID
    agent_id: UUID
    thread_id: UUID | None
    memory_type: MemoryType
    content: str
    importance: float
    access_count: int
    last_accessed_at: datetime | None
    expires_at: datetime | None
    metadata_json: dict
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


# ── Custom Agent Runs ────────────────────────────────────────────────────────

class CustomAgentRunResponse(BaseModel):
    id: UUID
    workspace_id: UUID
    agent_id: UUID
    thread_id: UUID | None
    status: str
    model_provider: str | None
    model_name: str | None
    input_json: dict
    output_json: dict
    error_json: dict
    tokens_input: int = 0
    tokens_output: int = 0
    cost_usd: float = 0
    latency_ms: int | None
    sub_agent_runs_json: list[dict]
    tool_call_count: int = 0
    knowledge_chunks_retrieved: int = 0
    metadata_json: dict
    started_at: datetime | None
    completed_at: datetime | None
    created_at: datetime
    steps: list["CustomAgentRunStepResponse"] = Field(default_factory=list)

    model_config = ConfigDict(from_attributes=True)


class CustomAgentRunStepResponse(BaseModel):
    id: UUID
    workspace_id: UUID
    run_id: UUID
    agent_name: str | None
    step_type: str
    step_order: int
    input_json: dict
    output_json: dict
    status: str
    error_message: str | None
    latency_ms: int | None
    created_at: datetime
    completed_at: datetime | None

    model_config = ConfigDict(from_attributes=True)


# ── Agent Templates ──────────────────────────────────────────────────────────

class AgentTemplateResponse(BaseModel):
    id: UUID
    name: str
    description: str
    category: str
    config_json: dict
    is_public: bool
    usage_count: int
    created_at: datetime

    model_config = ConfigDict(from_attributes=True)


# ── Agent Run Request ────────────────────────────────────────────────────────

class AgentRunRequest(BaseModel):
    thread_id: UUID | None = None
    message: str = Field(..., min_length=1, max_length=10000)
    metadata_json: dict = Field(default_factory=dict)


# ── Agent Chat Response (for streaming) ──────────────────────────────────────

class AgentChatResponse(BaseModel):
    run_id: UUID
    thread_id: UUID
    message: AgentMessageResponse | None = None
    status: str
    error: str | None = None


# ── Available Tools Catalog ──────────────────────────────────────────────────

class AvailableTool(BaseModel):
    tool_name: str
    display_name: str
    description: str
    category: str  # 'communication', 'development', 'data', 'web', 'business'
    default_permission_level: ToolPermissionLevel = ToolPermissionLevel.READ
    requires_approval_by_default: bool = False
    is_dangerous: bool = False


# ── Resolve forward references ───────────────────────────────────────────────

CustomAgentResponse.model_rebuild()
KnowledgeBaseResponse.model_rebuild()
CustomAgentRunResponse.model_rebuild()

File: apps/web/src/lib/api/custom-agents.ts
import { apiBaseUrl, defaultWorkspaceId, localAuthHeaders, readApiData } from "@/lib/api/client";

// ── Types ────────────────────────────────────────────────────────────────────

export type CustomAgentSummary = {
  id: string;
  workspace_id: string;
  name: string;
  description: string | null;
  model_provider: string;
  model_name: string;
  memory_enabled: boolean;
  status: "draft" | "active" | "paused" | "archived";
  created_at: string;
  updated_at: string;
  tool_count: number;
  sub_agent_count: number;
  thread_count: number;
};

export type CustomAgent = {
  id: string;
  workspace_id: string;
  name: string;
  description: string | null;
  role_instructions: string | null;
  model_provider: "openai" | "anthropic" | "deepseek" | "google" | "local";
  model_name: string;
  temperature: number;
  max_tokens: number;
  memory_enabled: boolean;
  memory_ttl_days: number;
  output_format: "text" | "json" | "markdown" | "table" | "code" | "email";
  output_schema: Record<string, unknown> | null;
  workflow_type: "sequential" | "parallel" | "supervisor" | "custom";
  workflow_graph: Record<string, unknown> | null;
  status: "draft" | "active" | "paused" | "archived";
  metadata_json: Record<string, unknown>;
  created_by: string | null;
  updated_by: string | null;
  created_at: string;
  updated_at: string;
  tools: AgentTool[];
  sub_agents: SubAgent[];
  knowledge_base_ids: string[];
};

export type AgentTool = {
  id: string;
  workspace_id: string;
  agent_id: string;
  tool_name: string;
  tool_config: Record<string, unknown>;
  enabled: boolean;
  permission_level: "read" | "write" | "execute" | "admin";
  requires_approval: boolean;
  rate_limit_per_day: number | null;
  allowed_domains: string[] | null;
  allowed_actions: string[] | null;
  created_at: string;
  updated_at: string;
};

export type AvailableTool = {
  tool_name: string;
  display_name: string;
  description: string;
  category: string;
  default_permission_level: "read" | "write" | "execute" | "admin";
  requires_approval_by_default: boolean;
  is_dangerous: boolean;
};

export type KnowledgeBase = {
  id: string;
  workspace_id: string;
  name: string;
  description: string | null;
  source_type: string;
  embedding_model: string;
  embedding_provider: string;
  file_count: number;
  chunk_count: number;
  status: "pending" | "processing" | "ready" | "error";
  created_by: string | null;
  created_at: string;
  updated_at: string;
  files: KnowledgeBaseFile[];
};

export type KnowledgeBaseFile = {
  id: string;
  knowledge_base_id: string;
  file_name: string;
  file_path: string;
  file_type: string;
  file_size_bytes: number | null;
  chunk_count: number;
  status: string;
  error_message: string | null;
  created_at: string;
};

export type SubAgent = {
  id: string;
  workspace_id: string;
  parent_agent_id: string;
  name: string;
  role: string;
  instructions: string;
  model_provider: string | null;
  model_name: string | null;
  tools_json: string[];
  execution_order: number;
  enabled: boolean;
  requires_approval: boolean;
  created_at: string;
  updated_at: string;
};

export type AgentThread = {
  id: string;
  workspace_id: string;
  agent_id: string;
  title: string | null;
  status: string;
  metadata_json: Record<string, unknown>;
  created_by: string | null;
  created_at: string;
  updated_at: string;
  message_count: number;
  last_message_at: string | null;
};

export type AgentMessage = {
  id: string;
  workspace_id: string;
  thread_id: string;
  agent_id: string | null;
  role: "user" | "assistant" | "system" | "tool" | "sub_agent";
  content: string | null;
  tool_calls_json: Record<string, unknown>[] | null;
  tool_call_id: string | null;
  sub_agent_name: string | null;
  token_count: number | null;
  metadata_json: Record<string, unknown>;
  created_at: string;
};

export type AgentMemory = {
  id: string;
  workspace_id: string;
  agent_id: string;
  thread_id: string | null;
  memory_type: "preference" | "fact" | "decision" | "context" | "user_info" | "project";
  content: string;
  importance: number;
  access_count: number;
  last_accessed_at: string | null;
  expires_at: string | null;
  metadata_json: Record<string, unknown>;
  created_at: string;
  updated_at: string;
};

export type AgentRun = {
  id: string;
  workspace_id: string;
  agent_id: string;
  thread_id: string | null;
  status: string;
  model_provider: string | null;
  model_name: string | null;
  input_json: Record<string, unknown>;
  output_json: Record<string, unknown>;
  error_json: Record<string, unknown>;
  tokens_input: number;
  tokens_output: number;
  cost_usd: number;
  latency_ms: number | null;
  sub_agent_runs_json: Record<string, unknown>[];
  tool_call_count: number;
  knowledge_chunks_retrieved: number;
  metadata_json: Record<string, unknown>;
  started_at: string | null;
  completed_at: string | null;
  created_at: string;
  steps: AgentRunStep[];
};

export type AgentRunStep = {
  id: string;
  workspace_id: string;
  run_id: string;
  agent_name: string | null;
  step_type: string;
  step_order: number;
  input_json: Record<string, unknown>;
  output_json: Record<string, unknown>;
  status: string;
  error_message: string | null;
  latency_ms: number | null;
  created_at: string;
  completed_at: string | null;
};

export type AgentTemplate = {
  id: string;
  name: string;
  description: string;
  category: string;
  config_json: Record<string, unknown>;
  is_public: boolean;
  usage_count: number;
  created_at: string;
};

// ── Tools Catalog ────────────────────────────────────────────────────────────

export async function getAvailableTools(workspaceId = defaultWorkspaceId): Promise<AvailableTool[]> {
  const response = await fetch(`${apiBaseUrl}/v1/workspaces/${workspaceId}/custom-agent-tools-catalog`, { headers: localAuthHeaders(workspaceId), cache: "no-store" });
  return readApiData<AvailableTool[]>(response, "Available tools could not be loaded.");
}

// ── Custom Agents CRUD ───────────────────────────────────────────────────────

export async function listCustomAgents(workspaceId = defaultWorkspaceId): Promise<CustomAgentSummary[]> {
  const response = await fetch(`${apiBaseUrl}/v1/workspaces/${workspaceId}/custom-agents`, { headers: localAuthHeaders(workspaceId), cache: "no-store" });
  return readApiData<CustomAgentSummary[]>(response, "Custom agents could not be loaded.");
}

export async function getCustomAgent(agentId: string, workspaceId = defaultWorkspaceId): Promise<CustomAgent> {
  const response = await fetch(`${apiBaseUrl}/v1/workspaces/${workspaceId}/custom-agents/${agentId}`, { headers: localAuthHeaders(workspaceId), cache: "no-store" });
  return readApiData<CustomAgent>(response, "Custom agent could not be loaded.");
}

export async function createCustomAgent(payload: Partial<CustomAgent> & { name: string }, workspaceId = defaultWorkspaceId): Promise<CustomAgent> {
  const response = await fetch(`${apiBaseUrl}/v1/workspaces/${workspaceId}/custom-agents`, {
    method: "POST",
    headers: { ...localAuthHeaders(workspaceId, "analyst"), "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });
  return readApiData<CustomAgent>(response, "Custom agent could not be created.");
}

export async function updateCustomAgent(agentId: string, payload: Partial<CustomAgent>, workspaceId = defaultWorkspaceId): Promise<CustomAgent> {
  const response = await fetch(`${apiBaseUrl}/v1/workspaces/${workspaceId}/custom-agents/${agentId}`, {
    method: "PATCH",
    headers: { ...localAuthHeaders(workspaceId, "analyst"), "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });
  return readApiData<CustomAgent>(response, "Custom agent could not be updated.");
}

export async function deleteCustomAgent(agentId: string, workspaceId = defaultWorkspaceId): Promise<{ id: string; deleted: boolean }> {
  const response = await fetch(`${apiBaseUrl}/v1/workspaces/${workspaceId}/custom-agents/${agentId}`, {
    method: "DELETE",
    headers: localAuthHeaders(workspaceId, "analyst"),
  });
  return readApiData<{ id: string; deleted: boolean }>(response, "Custom agent could not be deleted.");
}

// ── Agent Tools ──────────────────────────────────────────────────────────────

export async function listAgentTools(agentId: string, workspaceId = defaultWorkspaceId): Promise<AgentTool[]> {
  const response = await fetch(`${apiBaseUrl}/v1/workspaces/${workspaceId}/custom-agents/${agentId}/tools`, { headers: localAuthHeaders(workspaceId), cache: "no-store" });
  return readApiData<AgentTool[]>(response, "Agent tools could not be loaded.");
}

export async function addAgentTool(agentId: string, payload: Partial<AgentTool> & { tool_name: string }, workspaceId = defaultWorkspaceId): Promise<AgentTool> {
  const response = await fetch(`${apiBaseUrl}/v1/workspaces/${workspaceId}/custom-agents/${agentId}/tools`, {
    method: "POST",
    headers: { ...localAuthHeaders(workspaceId, "analyst"), "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });
  return readApiData<AgentTool>(response, "Tool could not be added.");
}

export async function updateAgentTool(agentId: string, toolId: string, payload: Partial<AgentTool>, workspaceId = defaultWorkspaceId): Promise<AgentTool> {
  const response = await fetch(`${apiBaseUrl}/v1/workspaces/${workspaceId}/custom-agents/${agentId}/tools/${toolId}`, {
    method: "PATCH",
    headers: { ...localAuthHeaders(workspaceId, "analyst"), "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });
  return readApiData<AgentTool>(response, "Tool could not be updated.");
}

export async function removeAgentTool(agentId: string, toolId: string, workspaceId = defaultWorkspaceId): Promise<{ id: string; deleted: boolean }> {
  const response = await fetch(`${apiBaseUrl}/v1/workspaces/${workspaceId}/custom-agents/${agentId}/tools/${toolId}`, {
    method: "DELETE",
    headers: localAuthHeaders(workspaceId, "analyst"),
  });
  return readApiData<{ id: string; deleted: boolean }>(response, "Tool could not be removed.");
}

// ── Knowledge Bases ──────────────────────────────────────────────────────────

export async function listKnowledgeBases(workspaceId = defaultWorkspaceId): Promise<KnowledgeBase[]> {
  const response = await fetch(`${apiBaseUrl}/v1/workspaces/${workspaceId}/knowledge-bases`, { headers: localAuthHeaders(workspaceId), cache: "no-store" });
  return readApiData<KnowledgeBase[]>(response, "Knowledge bases could not be loaded.");
}

export async function createKnowledgeBase(payload: { name: string; description?: string }, workspaceId = defaultWorkspaceId): Promise<KnowledgeBase> {
  const response = await fetch(`${apiBaseUrl}/v1/workspaces/${workspaceId}/knowledge-bases`, {
    method: "POST",
    headers: { ...localAuthHeaders(workspaceId, "analyst"), "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });
  return readApiData<KnowledgeBase>(response, "Knowledge base could not be created.");
}

export async function deleteKnowledgeBase(kbId: string, workspaceId = defaultWorkspaceId): Promise<{ id: string; deleted: boolean }> {
  const response = await fetch(`${apiBaseUrl}/v1/workspaces/${workspaceId}/knowledge-bases/${kbId}`, {
    method: "DELETE",
    headers: localAuthHeaders(workspaceId, "analyst"),
  });
  return readApiData<{ id: string; deleted: boolean }>(response, "Knowledge base could not be deleted.");
}

// ── Knowledge Base Links ─────────────────────────────────────────────────────

export async function linkKnowledgeBase(agentId: string, kbId: string, workspaceId = defaultWorkspaceId): Promise<{ agent_id: string; knowledge_base_id: string; linked: boolean }> {
  const response = await fetch(`${apiBaseUrl}/v1/workspaces/${workspaceId}/custom-agents/${agentId}/knowledge-bases/${kbId}`, {
    method: "POST",
    headers: localAuthHeaders(workspaceId, "analyst"),
  });
  return readApiData<{ agent_id: string; knowledge_base_id: string; linked: boolean }>(response, "Knowledge base could not be linked.");
}

export async function unlinkKnowledgeBase(agentId: string, kbId: string, workspaceId = defaultWorkspaceId): Promise<{ agent_id: string; knowledge_base_id: string; linked: boolean }> {
  const response = await fetch(`${apiBaseUrl}/v1/workspaces/${workspaceId}/custom-agents/${agentId}/knowledge-bases/${kbId}`, {
    method: "DELETE",
    headers: localAuthHeaders(workspaceId, "analyst"),
  });
  return readApiData<{ agent_id: string; knowledge_base_id: string; linked: boolean }>(response, "Knowledge base could not be unlinked.");
}

// ── Sub-Agents ───────────────────────────────────────────────────────────────

export async function listSubAgents(agentId: string, workspaceId = defaultWorkspaceId): Promise<SubAgent[]> {
  const response = await fetch(`${apiBaseUrl}/v1/workspaces/${workspaceId}/custom-agents/${agentId}/sub-agents`, { headers: localAuthHeaders(workspaceId), cache: "no-store" });
  return readApiData<SubAgent[]>(response, "Sub-agents could not be loaded.");
}

export async function createSubAgent(agentId: string, payload: { name: string; role: string; instructions: string; execution_order?: number }, workspaceId = defaultWorkspaceId): Promise<SubAgent> {
  const response = await fetch(`${apiBaseUrl}/v1/workspaces/${workspaceId}/custom-agents/${agentId}/sub-agents`, {
    method: "POST",
    headers: { ...localAuthHeaders(workspaceId, "analyst"), "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });
  return readApiData<SubAgent>(response, "Sub-agent could not be created.");
}

export async function updateSubAgent(agentId: string, subAgentId: string, payload: Partial<SubAgent>, workspaceId = defaultWorkspaceId): Promise<SubAgent> {
  const response = await fetch(`${apiBaseUrl}/v1/workspaces/${workspaceId}/custom-agents/${agentId}/sub-agents/${subAgentId}`, {
    method: "PATCH",
    headers: { ...localAuthHeaders(workspaceId, "analyst"), "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });
  return readApiData<SubAgent>(response, "Sub-agent could not be updated.");
}

export async function deleteSubAgent(agentId: string, subAgentId: string, workspaceId = defaultWorkspaceId): Promise<{ id: string; deleted: boolean }> {
  const response = await fetch(`${apiBaseUrl}/v1/workspaces/${workspaceId}/custom-agents/${agentId}/sub-agents/${subAgentId}`, {
    method: "DELETE",
    headers: localAuthHeaders(workspaceId, "analyst"),
  });
  return readApiData<{ id: string; deleted: boolean }>(response, "Sub-agent could not be deleted.");
}

// ── Threads & Messages ──────────────────────────────────────────────────────

export async function listAgentThreads(agentId: string, workspaceId = defaultWorkspaceId): Promise<AgentThread[]> {
  const response = await fetch(`${apiBaseUrl}/v1/workspaces/${workspaceId}/custom-agents/${agentId}/threads`, { headers: localAuthHeaders(workspaceId), cache: "no-store" });
  return readApiData<AgentThread[]>(response, "Threads could not be loaded.");
}

export async function createAgentThread(agentId: string, title?: string, workspaceId = defaultWorkspaceId): Promise<AgentThread> {
  const response = await fetch(`${apiBaseUrl}/v1/workspaces/${workspaceId}/custom-agents/${agentId}/threads`, {
    method: "POST",
    headers: { ...localAuthHeaders(workspaceId, "analyst"), "Content-Type": "application/json" },
    body: JSON.stringify({ agent_id: agentId, title: title || null }),
  });
  return readApiData<AgentThread>(response, "Thread could not be created.");
}

export async function listThreadMessages(threadId: string, workspaceId = defaultWorkspaceId): Promise<AgentMessage[]> {
  const response = await fetch(`${apiBaseUrl}/v1/workspaces/${workspaceId}/threads/${threadId}/messages`, { headers: localAuthHeaders(workspaceId), cache: "no-store" });
  return readApiData<AgentMessage[]>(response, "Messages could not be loaded.");
}

export async function runAgent(agentId: string, message: string, threadId?: string, workspaceId = defaultWorkspaceId): Promise<AgentRun> {
  const response = await fetch(`${apiBaseUrl}/v1/workspaces/${workspaceId}/custom-agents/${agentId}/run`, {
    method: "POST",
    headers: { ...localAuthHeaders(workspaceId, "analyst"), "Content-Type": "application/json" },
    body: JSON.stringify({ message, thread_id: threadId || null }),
  });
  return readApiData<AgentRun>(response, "Agent run could not be started.");
}

export async function getAgentRun(runId: string, workspaceId = defaultWorkspaceId): Promise<AgentRun> {
  const response = await fetch(`${apiBaseUrl}/v1/workspaces/${workspaceId}/custom-agent-runs/${runId}`, { headers: localAuthHeaders(workspaceId), cache: "no-store" });
  return readApiData<AgentRun>(response, "Agent run could not be loaded.");
}

// ── Memories ────────────────────────────────────────────────────────────────

export async function listAgentMemories(agentId: string, workspaceId = defaultWorkspaceId): Promise<AgentMemory[]> {
  const response = await fetch(`${apiBaseUrl}/v1/workspaces/${workspaceId}/custom-agents/${agentId}/memories`, { headers: localAuthHeaders(workspaceId), cache: "no-store" });
  return readApiData<AgentMemory[]>(response, "Memories could not be loaded.");
}

export async function createAgentMemory(agentId: string, payload: { content: string; memory_type: string; importance?: number }, workspaceId = defaultWorkspaceId): Promise<AgentMemory> {
  const response = await fetch(`${apiBaseUrl}/v1/workspaces/${workspaceId}/custom-agents/${agentId}/memories`, {
    method: "POST",
    headers: { ...localAuthHeaders(workspaceId, "analyst"), "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });
  return readApiData<AgentMemory>(response, "Memory could not be created.");
}

export async function deleteAgentMemory(agentId: string, memoryId: string, workspaceId = defaultWorkspaceId): Promise<{ id: string; deleted: boolean }> {
  const response = await fetch(`${apiBaseUrl}/v1/workspaces/${workspaceId}/custom-agents/${agentId}/memories/${memoryId}`, {
    method: "DELETE",
    headers: localAuthHeaders(workspaceId, "analyst"),
  });
  return readApiData<{ id: string; deleted: boolean }>(response, "Memory could not be deleted.");
}

// ── Templates ───────────────────────────────────────────────────────────────

export async function listAgentTemplates(workspaceId = defaultWorkspaceId): Promise<AgentTemplate[]> {
  const response = await fetch(`${apiBaseUrl}/v1/workspaces/${workspaceId}/agent-templates`, { headers: localAuthHeaders(workspaceId), cache: "no-store" });
  return readApiData<AgentTemplate[]>(response, "Templates could not be loaded.");
}

export async function cloneAgentTemplate(templateId: string, workspaceId = defaultWorkspaceId): Promise<CustomAgent> {
  const response = await fetch(`${apiBaseUrl}/v1/workspaces/${workspaceId}/agent-templates/${templateId}/clone`, {
    method: "POST",
    headers: localAuthHeaders(workspaceId, "analyst"),
  });
  return readApiData<CustomAgent>(response, "Template could not be cloned.");
}